File: projects/frontend/src/lib/onboardingCompanyName.ts
/**
 * Industry-themed word lists for the first part of generated company names.
 * Each industry has 10+ words for sufficient variety across 10+ regenerations.
 */
const industryWords: Record<string, string[]> = {
  FURNITURE: ['Oak', 'Timber', 'Cedar', 'Maple', 'Walnut', 'Birch', 'Crafted', 'Pine', 'Teak', 'Redwood', 'Ironwood', 'Ember'],
  FOOD_PROCESSING: ['Harvest', 'Golden', 'Granary', 'Valley', 'Meadow', 'Artisan', 'Heritage', 'Fresh', 'Prime', 'Grain', 'Sunrise', 'Orchard'],
  HEALTHCARE: ['Vital', 'Nova', 'Remedy', 'Shield', 'Helix', 'Apex', 'Synapse', 'BioNex', 'Medus', 'Curative', 'Lumina', 'ClareMed'],
  ELECTRONICS: ['Circuit', 'Silicon', 'Quantum', 'Pixel', 'Axion', 'Pulse', 'Syntek', 'Nano', 'Logic', 'Nexon', 'CoreTech', 'Voltaic'],
  CONSTRUCTION: ['Bedrock', 'Summit', 'Keystone', 'Ironclad', 'Pillar', 'Arcstone', 'Granite', 'Crest', 'Foundry', 'Basalt', 'Vantage', 'Stronghold'],
  PHARMACEUTICALS: ['Helix', 'BioNex', 'Vita', 'Novex', 'Synapse', 'Curative', 'Shield', 'Medic', 'Lumena', 'Apharma', 'Zoria', 'GenoPlex'],
  ENERGY: ['Solaris', 'Voltan', 'Kinetic', 'Radiant', 'Dynamo', 'Fusion', 'Horizon', 'PowerCo', 'GridTech', 'Wattex', 'Ember', 'Voltaic'],
  LOGISTICS: ['Meridian', 'Nexus', 'Convoy', 'Pathfin', 'Cargotek', 'Swift', 'TransMax', 'Linker', 'Freightco', 'Relay', 'FluxNet', 'Harborcroft'],
}

/** Business-type suffixes shared across all industries. */
const businessSuffixes = [
  'Industries',
  'Ventures',
  'Capital',
  'Works',
  'Corp',
  'Solutions',
  'Group',
  'Holdings',
  'Dynamics',
  'Partners',
  'Collective',
  'House',
]

function hashSeed(value: string): number {
  let hash = 0
  for (const char of value) {
    hash = (hash * 31 + char.charCodeAt(0)) >>> 0
  }
  return hash
}

/**
 * Generates a professional two-word company name suitable for an economic simulation game.
 *
 * @param industry - The selected industry key (e.g. 'FURNITURE', 'HEALTHCARE').
 * @param cityName - Optional city name used to diversify the base seed.
 * @param offset   - Regeneration counter; increment to get a different name for the same inputs.
 *                   Defaults to 0 (first suggestion). Values 0–11 always produce distinct names.
 */
export function generateOnboardingCompanyName(industry: string, cityName?: string | null, offset = 0): string {
  const normalizedIndustry = industry || 'GENERAL'
  const normalizedCity = cityName?.trim() || 'Capital'
  const baseSeed = hashSeed(`${normalizedIndustry}:${normalizedCity}`)

  const words = industryWords[normalizedIndustry] ?? ['Prime', 'Atlas', 'Summit', 'Nexus', 'Apex', 'Vanguard', 'Pinnacle', 'Core', 'Axion', 'Titan', 'Zenith', 'Crest']

  // Use the offset to step through both lists independently so each regeneration gives
  // a fresh combination and all businessSuffixes are eventually used.
  const wordIndex = (baseSeed + offset) % words.length
  const suffixIndex = (baseSeed + offset * 3 + 7) % businessSuffixes.length

  const word = words[wordIndex] ?? words[0]!
  const suffix = businessSuffixes[suffixIndex] ?? businessSuffixes[0]!

  return `${word} ${suffix}`
}

/** Total number of distinct names the generator can produce for any single industry/city pair. */
export const NAME_GENERATOR_CYCLE_LENGTH = businessSuffixes.length

File: projects/flutter_app/lib/features/onboarding/onboarding_company_name.dart
// Ported from `projects/frontend/src/lib/onboardingCompanyName.ts`. Web uses
// the `unique-names-generator` package purely as a "pick one of dict A + one
// of dict B" combiner — reproduced here with `Random` directly, no package
// needed for that part.

import 'dart:math';

/// Industry-themed first-word lists, mirroring web's `industryWords` exactly.
const Map<String, List<String>> industryWords = {
  'FURNITURE': [
    'Oak', 'Timber', 'Cedar', 'Maple', 'Walnut', 'Birch', 'Pine', 'Teak',
    'Redwood', 'Ironwood', 'Crafted', 'Craftwood', 'Laurelwood', 'Classic', 'Cherrywood',
    'Ashwood', 'Elmwood', 'Mahogany', 'Rosewood', 'Sandalwood',
  ],
  'FOOD_PROCESSING': [
    'Harvest', 'Golden', 'Granary', 'Valley', 'Meadow', 'Brewer', 'Savorex',
    'Fresh', 'Prime', 'Grain', 'Sunrise', 'Orchard', 'Bloom', 'Pastoral',
    'Crofton', 'Millstone', 'Croftsman', 'Barley', 'Amber', 'Verdant',
  ],
  'HEALTHCARE': [
    'Vital', 'Nova', 'Remedy', 'Shield', 'Helix', 'Apex', 'Synapse',
    'Curative', 'Lumina', 'Medic', 'Regen', 'Clarity', 'Salus', 'Biovance',
    'Correx', 'Nuvarix', 'Lifespan', 'Vitalix', 'Pharmax', 'Zenova',
  ],
  'ELECTRONICS': [
    'Circuit', 'Silicon', 'Quantum', 'Pixel', 'Pulse', 'Nano', 'Logic',
    'Voltaic', 'CoreTech', 'Axiom', 'Nexon', 'Syntek', 'Byte', 'Lattice',
    'Photon', 'Hexagon', 'Vertex', 'Microcore', 'Orbis', 'Telaris',
  ],
  'CONSTRUCTION': [
    'Bedrock', 'Summit', 'Keystone', 'Ironclad', 'Pillar', 'Granite',
    'Crest', 'Foundry', 'Basalt', 'Bulwark', 'Stronghold', 'Arcstone',
    'Trident', 'Rampart', 'Citadel', 'Solida', 'Terrafirm', 'Pinnacle',
    'Masonry', 'Cornerstone',
  ],
  'PHARMACEUTICALS': [
    'Scienta', 'Vita', 'Novex', 'Biospan', 'Pharmex', 'Clineva', 'Lumena',
    'Apharma', 'Zoria', 'GenoPlex', 'Theravo', 'Biovanta', 'Cellex',
    'Therapeutix', 'Vitalora', 'Genexis', 'Pharmasol', 'Clinex', 'Medivance',
    'Revitare',
  ],
  'ENERGY': [
    'Solaris', 'Kinetic', 'Radiant', 'Dynamo', 'Fusion', 'Horizon', 'Ignition',
    'Wattcore', 'Wattex', 'GridTech', 'PowerCo', 'Voltan', 'Lumex',
    'Ampere', 'Therma', 'Photona', 'Polaris', 'Celero', 'Novagen', 'Electra',
  ],
  'LOGISTICS': [
    'Meridian', 'Nexus', 'Convoy', 'Swift', 'Relay', 'FluxNet', 'TransMax',
    'Harborcroft', 'Cargotek', 'Linker', 'Freightco', 'Pathfinder', 'Vectus',
    'Transito', 'Velocity', 'Expedio', 'Bridgeport', 'Portside', 'Xpedite',
    'Trailblazer',
  ],
};

/// Fallback word list for unknown/unmapped industries, mirroring web's `fallbackWords`.
const List<String> fallbackWords = [
  'Prime', 'Atlas', 'Summit', 'Nexus', 'Apex', 'Vanguard', 'Pinnacle',
  'Core', 'Titan', 'Zenith', 'Crest', 'Orion', 'Solace', 'Fortis',
  'Verdant', 'Axiom', 'Crestline', 'Triton', 'Halcyon', 'Meridian',
];

/// Shared business-type suffixes, mirroring web's `businessSuffixes`.
const List<String> businessSuffixes = [
  'Industries', 'Ventures', 'Capital', 'Works', 'Corp', 'Solutions', 'Group',
  'Holdings', 'Dynamics', 'Partners', 'Collective', 'House', 'Enterprises',
  'Trading', 'Global',
];

String _sessionKey = '';
final Set<String> _usedCompanyNames = {};

/// Resets the in-session uniqueness tracker when the player changes industry
/// or city, so the next [generateOnboardingCompanyName] call starts fresh.
void resetCompanyNameSession(String key) {
  if (key != _sessionKey) {
    _sessionKey = key;
    _usedCompanyNames.clear();
  }
}

/// Generates a "Word Suffix" company name for [industry], retrying up to 50
/// times to avoid repeating a name already shown in this session (cleared via
/// [resetCompanyNameSession]). Mirrors `generateOnboardingCompanyName`.
String generateOnboardingCompanyName(String industry, {Random? random}) {
  final rand = random ?? Random();
  final words = industryWords[industry] ?? fallbackWords;

  String name = '';
  for (var i = 0; i < 50; i++) {
    name = '${words[rand.nextInt(words.length)]} ${businessSuffixes[rand.nextInt(businessSuffixes.length)]}';
    if (!_usedCompanyNames.contains(name)) break;
  }

  if (_usedCompanyNames.contains(name)) {
    _usedCompanyNames.clear();
    name = '${words[rand.nextInt(words.length)]} ${businessSuffixes[rand.nextInt(businessSuffixes.length)]}';
  }

  _usedCompanyNames.add(name);
  return name;
}
